src/operators.jl

```diff
@@ -17,24 +17,41 @@ associative, and if `iscommutative` is true, assumed to be commutative as well.
 - [`Scan!`](@ref)/[`Scan`](@ref)
 - [`Exscan!`](@ref)/[`Exscan`](@ref)
 """
-@mpi_handle Op MPI_Op fptr
-
-const OP_NULL = _Op(MPI_OP_NULL, nothing)
-const BAND    = _Op(MPI_BAND, nothing)
-const BOR     = _Op(MPI_BOR, nothing)
-const BXOR    = _Op(MPI_BXOR, nothing)
-const LAND    = _Op(MPI_LAND, nothing)
-const LOR     = _Op(MPI_LOR, nothing)
-const LXOR    = _Op(MPI_LXOR, nothing)
-const MAX     = _Op(MPI_MAX, nothing)
-const MIN     = _Op(MPI_MIN, nothing)
-const PROD    = _Op(MPI_PROD, nothing)
-const REPLACE = _Op(MPI_REPLACE, nothing)
-const SUM     = _Op(MPI_SUM, nothing)
-
-if @isdefined(MPI_NO_OP)
-    const NO_OP   = _Op(MPI_NO_OP, nothing)
+mutable struct Op
+    val::MPI_Op
+    fptr
+    Op(val::MPI_Op, fptr) = new(val, fptr)
 end
+Base.:(==)(a::Op, b::Op) = a.val == b.val
+Base.cconvert(::Type{MPI_Op}, op::Op) = op.val
```

Comment:
This should get rid of the need for `GC.@preserve` in `MPI.Reduce!`
```suggestion
Base.cconvert(::Type{MPI_Op}, op::Op) = opl
Base.unsafe_convert(::Type{MPI_Op}, op::Op) = op.val
```

Reply:
Can you explain? Are you saying that `cconvert` is called before `GC_PUSH`, while `unsafe_convert` is called after?

Reply:
That's my understanding: in the docs for [`unsafe_convert`](https://docs.julialang.org/en/v1/base/c/#Base.unsafe_convert):
> In cases where convert would need to take a Julia object and turn it into a Ptr, this function should be used to define and perform that conversion.

Reply:
(We're not converting to a pointer. `MPI_Op` might be a typedef'd `int` in C.)

The documentation also says `Be careful to ensure that a Julia reference to x exists as long as the result of this function will be used.` I think that's the condition that is violated here:
```Julia
function Reduce!(rbuf::RBuffer, op::Union{Op,MPI_Op}, root::Integer, comm::Comm)
    @mpichk ccall((:MPI_Reduce, libmpi), Cint,
                  (MPIPtr, MPIPtr, Cint, MPI_Datatype, MPI_Op, Cint, MPI_Comm),
                  rbuf.senddata, rbuf.recvdata, rbuf.count, rbuf.datatype, op, root, comm)
    return rbuf.recvdata
end
```
The argument `op` might be passed in a register, and since `op` is not used after passing it to `MPI_Reduce`, the register might be used otherwise (e.g. holding `op.val`). Thus there is no live reference left during the call to `MPI_Reduce`.

Reply:
The input to `unsafe_convert` is rooted.

```
julia> function f(x)
          ccall(:jl_, Cvoid, (Ptr{Int},), x)
       end
f (generic function with 1 method)

julia> @code_llvm optimize=false f(Ref(10))
;  @ REPL[1]:1 within `f`
define nonnull {}* @japi1_f_126({}* %0, {}** %1, i32 %2) #0 {
top:
  %3 = alloca {}**, align 8
  store volatile {}** %1, {}*** %3, align 8
  %4 = call {}*** @julia.get_pgcstack()
  %5 = bitcast {}*** %4 to {}**
  %current_task = getelementptr inbounds {}*, {}** %5, i64 2305843009213693940
  %6 = bitcast {}** %current_task to i64*
  %world_age = getelementptr inbounds i64, i64* %6, i64 13
  %7 = getelementptr inbounds {}*, {}** %1, i64 0
  %8 = load {}*, {}** %7, align 8
;  @ REPL[1]:2 within `f`
; ┌ @ refvalue.jl:40 within `unsafe_convert`
; │┌ @ pointer.jl:147 within `pointer_from_objref`
    %9 = call nonnull {}* @julia.pointer_from_objref({}* %8) #1
    %10 = ptrtoint {}* %9 to i64
; └└
  call void inttoptr (i64 139891689567888 to void (i64)*)(i64 %10) [ "jl_roots"({}* %8) ]
  ret {}* inttoptr (i64 139891598184456 to {}*)
}

```

See the `jl_roots` operand bundle on the call.

Reply:
Done.